### ISSUE TITLE
Inconsistencies in COMPLETED/FAILED times and logs

### ISSUE THREAD

**Issue by LuiggiTenorioK** [BSC-ES/autosubmit-gui#113]
In GitLab by @mcastril on Jul 9, 2024, 16:22

This has implications in the log visualization, the correct identification of runs, and also in the performance metrics due to some jobs reporting unrealistic metrics that force us to implement outlier removal algorithms.

I take experiment `a0uo` as example:

![Screenshot_from_2024-07-09_16-12-03](https://earth.bsc.es/gitlab/es/autosubmit-gui/uploads/00a1b9a348009db3d1974034b5031efc/Screenshot_from_2024-07-09_16-12-03.png)

The .err and .out logs for the last run, according to the GUI, point to logs from different runs. Maybe the fact that the latest .out has been compressed is the cause since it's not findable by the GUI/API. If that's the case it should be more robust and not point to the latest available, but for the one that really corresponds to the run.

Then, in the job history view, we see a job COMPLETED in only 30 seconds.

![Screenshot_from_2024-07-09_16-14-25](https://earth.bsc.es/gitlab/es/autosubmit-gui/uploads/cd450e0c04309ded0a86ccd76ceb46b7/Screenshot_from_2024-07-09_16-14-25.png)

The job COMPLETED in 8:43:53 supposedly ran later, but this is not the one shown in the tree view when we open it:

![Screenshot_from_2024-07-09_16-18-32](https://earth.bsc.es/gitlab/es/autosubmit-gui/uploads/d65b93c2c05a8003d94ab228f1b27342/Screenshot_from_2024-07-09_16-18-32.png)

If we open the `/appl/AS/AUTOSUBMIT_DATA/a0uo/tmp/a0uo_19900101_fc0_11_SIM_TOTAL_STATS` file, the FAILED jobs don't have an ending time (then I infer this information is added in the DDBB directly by Autosubmit when the job is failed). The COMPLETED job ran for 8:34h, like the one with counter 62 shown above. There is no trail about the job running in 30 seconds.

```
20240301223744 20240301223905 19700101020000 FAILED
20240301233222 20240302044603 19700101020000 FAILED
20240302044634 20240302045337 19700101020000 FAILED
20240302045408 20240302045752 19700101020000 FAILED
20240302131902 20240302132206 19700101020000 FAILED
20240302151731 20240302152345 19700101020000 FAILED
20240302152416 20240302152614 20240303001007 COMPLETED
```
Unfortunately, there is no AS log for this run, and I cannot look there at what could happen. But this behaviour is repeated in other experiments and could be tracked by someone. 

Feel free to move the part of this issue that corresponds to Autosubmit to the other project. Sorry for open a single issue for the whole problem: @LuiggiTenorioK @dbeltrankyl @kinow

**Comment by LuiggiTenorioK** [comment]
What seems strange about this inconsistency is the fact that the Tree View shows the second to last execution of the job. It uses the Pkl information which is expected to have the latest info of the job but this is not the case.

Then, the job history table uses the `job_data_a0uo.db` DDBB to show the runs of that job which now seems to be the complete data.

And finally, we have the `a0uo_19900101_fc0_11_SIM_TOTAL_STATS` log that has just one of the two COMPLETE jobs and, also, has one FAILED job more than the job history table.

We have to review how data is written on Autosubmit to see what is wrong :sweat: 

Is it possible that the pkl stops updating after the first `COMPLETE` but there were remaining processes that continue updating the two other sources of data?

**Comment by LuiggiTenorioK** [comment]
In GitLab by @mcastril on Jul 10, 2024, 09:57

Thanks Luiggi

Then, also the fact that the job running for 30s is marked as COMPLETED. This could be due to a `setstatus`, but we are not sure.

Actually, the second-to-last job marked in TOTAL_STATS as FAILED is the one shown in the Tree view as COMPLETED. The starting time is 15:23:45, I guess UTC?

Then the `job_data_a0uo.db` could be complete but inconsistent, if it's showing the second-to-last as COMPLETED.

So, we have:
* Failed jobs marked as COMPLETED (in the GUI)
* Tree view showing not the last but the second-to-last run
* Incomplete pkl (probably related)
* Logs pointing to files belonging to different executions

This is important to debug because users may lose confidence in the tool if they see we tag failed jobs as completed, and also because it's impacting the performance metrics

**Comment by LuiggiTenorioK** [comment]
In GitLab by @dbeltrankyl on Jul 10, 2024, 11:36

Hello @mcastril , @LuiggiTenorioK 

Do you know which version this experiment was used? I see that in the minimal this was generated with 4.0.104

We had some fixes in v4.1.1 but also in v4.1.4 https://earth.bsc.es/gitlab/es/autosubmit/-/blob/master/CHANGELOG so unless they use 4.1.8 I think there could be some issues that we already debugged and fixed

**Comment by LuiggiTenorioK** [comment]
In GitLab by @mcastril on Jul 10, 2024, 12:26

Hi @dbeltrankyl, I took one random example, but this behavior is happening in quite a lot of experiments (that's why the outlier detection feature was implemented in the Performance Metrics table).

You can take a lot here: https://earth.bsc.es/gitlab/es/autosubmit-api/-/issues/83

So, on the one hand it's relieving that you think that the current Autosubmit version should not have these problems. On the other hand, I feel that if when the issue appears, we don't pay enough attention to it because experiments are not running with the latest version, we have the risk of overseeing it.

We must have the certainty that this is not happening because, as I said, we are reporting this information in Destination Earth (included in the Performance Metrics), and also, this tool is used by multiple users that may not report these problems but can get disappointed if we are not reporting accurate workflow information.

If you suggest so, you could also check the experiments running with v4.1.8 and verify that this behavior is not reproduced in any of those. I am looking for these issues from time to time, but I cannot do them as frequently as I can. However, I think this is an exercise that should be done routinely (at least until test coverage increases), to check the sanity of the data and what users are seeing.

**Comment by LuiggiTenorioK** [comment]
In GitLab by @mcastril on Aug 30, 2024, 11:41

Kai reported another inconsistency

The 79th chunk for the `a1bc` experiment running in DestinE had a failure and is marked as COMPLETED

![Screenshot_from_2024-08-30_11-29-30](https://earth.bsc.es/gitlab/es/autosubmit-gui/uploads/dcbc2cb7b102360ea961740570bdb495/Screenshot_from_2024-08-30_11-29-30.png)

As you can see in the plot, the time is not realistic, compared to others

The history shows that there was the following COMPLETED job, whose timing is also strange because it takes 3 times more than the average:

![Screenshot_from_2024-08-30_11-38-31](https://earth.bsc.es/gitlab/es/autosubmit-gui/uploads/6e1676ecc872a29e65a0ca73b177ff84/Screenshot_from_2024-08-30_11-38-31.png)

This experiment is using Autosubmit v4.1.9 @dbeltrankyl

https://climatedt-wf.csc.fi/experiment/a1bc/tree

**Comment by LuiggiTenorioK** [comment]
In GitLab by @mcastril on Aug 30, 2024, 11:44

The chunk 59 of the same experiment has another issue. This one doesn't impact the performance view because it's an extreme case, and it's filtered out. What I see here is an inconsistency in the data showed in the details box and in the history.

![Screenshot_from_2024-08-30_11-43-05](https://earth.bsc.es/gitlab/es/autosubmit-gui/uploads/4bba470df77533440adbcc62507c7324/Screenshot_from_2024-08-30_11-43-05.png)

This one is also wrongly marked as COMPLETED @dbeltrankyl 

What is also strange is that the SYPD and ASPYD values in the history box are incorrect. It has the SYPD and ASYPD set as it would have run for 6-7 hours, as the jobs that were actually completed. Sounds like it's taking the value from another instance? @LuiggiTenorioK

**Comment by LuiggiTenorioK** [comment]
> What is also strange is that the SYPD and ASPYD values in the history box are incorrect. It has the SYPD and ASYPD set as it would have run for 6-7 hours, as the jobs that were actually completed. Sounds like it's taking the value from another instance? @LuiggiTenorioK 

Hi Miguel, what it seems that it's happening its that there are different chunk units/sizes between runs. 

In this case, the run 22 had a chunk size 1 month which makes the SYPD be 1.61 in the job history but in the panel it takes the current chunk size of the YML which is 6 months resulting to a SYPD of 9.64.

**Comment by LuiggiTenorioK** [comment]
In GitLab by @dbeltrankyl on Aug 30, 2024, 12:35

Thanks for the report. I'll check it

**Comment by LuiggiTenorioK** [comment]
In GitLab by @dbeltrankyl on Aug 30, 2024, 12:36

mentioned in merge request autosubmit!475

**Comment by LuiggiTenorioK** [comment]
In GitLab by @mcastril on Aug 30, 2024, 12:37

That's a good point Luiggi. I think it should be consistent and show the same information. Would it be possible to correct it so that it shows the correct SYPD? Including also the Performance view.

I know this is not easy and can be complex, and may have implications, so would require to open another issue.

**Comment by LuiggiTenorioK** [comment]
Yes, we have to create an issue in the API fixing what is shown in the bottom-right panel (and also in the Performance view).

However, will it be weird that in one experiment different jobs have different chunk sizes(units) when they are supposed to have just a single consistent chunk unit/size?

**Comment by LuiggiTenorioK** [comment]
On the other hand, which is actually the right value? 

From my perspective, they are both valid. The job history shows the SYPD calculated based on the chunk definition at the time of its submission and the right-bottom panel shows it with the current chunk definition. Maybe it could be helpful to keep both values with a tooltip that explains why the difference.

**Comment by LuiggiTenorioK** [comment]
In GitLab by @mcastril on Aug 30, 2024, 13:16

> However, will it be weird that in one experiment different jobs have different chunk sizes(units) when they are supposed to have just a single consistent chunk unit/size?

It can happen. It's not usual but sometimes users are forced to do this because the jobs could start experiencing issues that can be mitigated by shortening the length

> On the other hand, which is actually the right value?

In my opinion, the correct one is the one that calculates the metric with the values on the submission time.

The chunk size gives us an estimation of how much progress the simulation made in one step, and we use the value to calculate the SYPD/ASYPD. If the chunksize is wrong we are not calculating well the SYPD / ASYPD

This is related with the logging of the configuration too, that we should ideally track everytime that it changes and not only a few times. I had pending to create an issue for that

**Comment by LuiggiTenorioK** [comment]
mentioned in issue autosubmit-api#86

**Comment by LuiggiTenorioK** [comment]
In GitLab by @mcastril on Sep 3, 2024, 10:25

I assigned the issue to you Dani

**Comment by LuiggiTenorioK** [comment]
In GitLab by @mcastril on Sep 17, 2024, 13:03

@dbeltrankyl could you make any progress here? I understand that you may be focused in finishing the fix for the zombies. If that's the case, please take this as the most important after that one was done. The consistency of the DDBB is quite important

**Comment by LuiggiTenorioK** [comment]
In GitLab by @dbeltrankyl on Sep 17, 2024, 13:41

Hello @mcastril ,

Yes, actually, I have been working on this in that merge request as the functions are inside the log recovery function. ( although with EDITO I didn't have enough time). I've fixed the issue there and wrote a regression test that found some issues with the retrial jobs timestamps.

The double entry was a mistake on my side. I had a function called twice by mistake.

However, I found other issues, and to fix them, I had to modify some functions that Wilmer created.

These finds are with unwrapped jobs.

Now, I believe that all the issues have been solved there, and I'm working on writing the same tests but for vertical wrappers, as the mechanism is a little different.

So, to be able to test for vertical wrappers, I had to enable the vertical wrappers for the local platform so I don't have to submit anything to mn5. 

I just managed to launch a wrapper with the local platform and now I'm writing the tests for the vertical wrapper.

So, in summary, with the addition of the database tests, we will be able to easily monitor/fix the job_data.db and be more reliable in the future

**Comment by LuiggiTenorioK** [comment]
In GitLab by @dbeltrankyl on Sep 17, 2024, 13:43

The branch is there https://earth.bsc.es/gitlab/es/autosubmit/-/commits/!475-vertical-regression-test, that will be merged into the zombies one once the tests are working and after that I'll ask somebody to review the branch

**Comment by LuiggiTenorioK** [comment]
In GitLab by @mcastril on Sep 17, 2024, 13:55

Thank you Dani. The addition of the tests is quite relevant. It's important to detect regressions before they are propagated

**Comment by LuiggiTenorioK** [comment]
In GitLab by @dbeltrankyl on Sep 26, 2024, 09:58

Hello, 

Finally, I managed to get all dates to work correctly in both the job_data and total_stats files. It was more complicated than expected.

I'll comment on the branch where I did the [changes](https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/490)  (will comment in https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/475), but in summary:

Had to :

* Implement Local wrappers
* Change vertical-wrapper cmd
* Change the way of updating the db
* Change the way of retrieving the row in the db
* Change the way of updating the local log variable
* Change the way of writing the different timestamps
* Change the way of detecting a failed wrapper and inner job.. but this is only for local platform as the slurm was detecting this okay
* Write four regression tests so it doesn't happen again, which will also greatly help me debug.

What could still be inconsistent:

It is the END date when a job finishes unsuccessfully and is *not* under a vertical wrapper.

Why? 

Because autosubmit fills this date when it detects that the job has failed. This can be a few seconds of error, so it is not a great issue  but

We can more or less solve it if:
 
1) Get the mod date of the .err in the remote_platform and make a timestamp of the last modification date
2) Trap all exits of the template, (I think there was an issue for that @kinow ?), and if the exit is trapped, we can easily append a failure timestamp to the STAT file (not sure how to trap the bash template tho) 

I didn't do in that branch because there is already a lot of changes and these need to touch even more code

**Comment by kinow** [comment]
Note: I checked my testing suite jobs launched with AS 4.1.12 latest in VM today, and I was not able to reproduce this issue -- could be that my experiments need wrappers or some other setting to do so, but just to report that I haven't seen this exact issue happening (I did see one where entire jobs of previous runs were available in the GUI though, which may (maybe?) be the same issue here, or related - https://github.com/BSC-ES/autosubmit/issues/1485
